pandas/_libs/tslibs/offsets.pyx

```diff
@@ -375,3 +375,33 @@ class BaseOffset(_BaseOffset):
             # i.e. isinstance(other, (ABCDatetimeIndex, ABCSeries))
             return other - self
         return -self + other
+
+
+# ----------------------------------------------------------------------
+# RelativeDelta Arithmetic
+
+
+cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
+    cdef:
+        int year, month, day
+        int dim, dy
+
+    dy = (stamp.month + months) // 12
+    month = (stamp.month + months) % 12
+
+    if month == 0:
+        month = 12
+        dy -= 1
+    year = stamp.year + dy
+
+    dim = monthrange(year, month)[1]
+    if day_opt is None:
+        day = min(stamp.day, dim)
+    elif day_opt == 'start':
+        day = 1
+    elif day_opt == 'end':
+        day = dim
+    else:
+        # assume this is an integer (and a valid day)
```

Comment:
? why is day_opt anything else?

Reply:
and if it is, then explicity put it.

Reply:
>? why is day_opt anything else?
 
For e.g. semi-month offsets we may be shifting to a particular day other than the first or last of the month.

> and if it is, then explicity put it.

You mean assert it?  OK.  I'll go ahead and write a docstring too.

Reply:
check if it’s an integer

the else clause should raise